### DISCUSSION TITLE
It seems that some java jdtls settings such as java.autobuild.enabled=false and java.jdt.ls.vmargs has not been supported?

### DISCUSSION THREAD

**Discussion by huangfeiyu** [joaotavora/eglot#1417] [Q&A] [answered]
Hi, first, let me thank you for your great work!
I have a hard time to set up jdtls for java development properly. I want to disable the autobuild also pass additional vmargs to jdtls, and I don't succeed. The following is my configuration, java.configuration.runtimes works, bundles works too. But not java.autobuild.enabled. Any suggestion? Thanks!
``` lisp
(with-eval-after-load 'eglot
  (add-to-list 'eglot-server-programs
               `(java-mode . ("jdtls" :initializationOptions
                              (
                               :settings
                               (
                                :java
                                (
                                 :autobuild (:enabled :json-false) ;; nil doesn't work either.
                                 :configuration (:runtimes [(:name "JavaSE-11" :path "/usr/local/jdk-11.0.21+9" :default t)
                                                            (:name "JavaSE-17" :path "/usr/lib/jvm/java-17-openjdk-amd64")])
                                 )
                                )
                               :bundles ["/usr/local/jdtls/com.microsoft.java.debug.plugin-0.51.1.jar"]
                               )
                              )))

  
  )
```

**Comment by yveszoundi** [comment]
For custom JVM args, you might want to provide a custom function to initialize `jdtls`, as described here: https://github.com/joaotavora/eglot/discussions/1185#discussioncomment-5346957 (`jdtls-command-contact` function). You'll then need to update `eglot-server-programs` accordingly.

For `autobuild` settings, your syntax seems correct at first sight. If I were you, I would double-check the eglot logs to see what configuration gets sent to the server or if there are any Java stacktraces from Eclipse (eglot events buffer).

**Reply by huangfeiyu** [reply]
@yveszoundi Thanks! The following configuration gets sent to the server according to eglot events buffer
``` json
{
  "initializationOptions": {
    "settings": {
      "java": {
        "autobuild": {
          "enabled": false
        },
        "configuration": {
          "runtimes": [
            {
              "name": "JavaSE-11",
              "path": "/usr/local/jdk-11.0.21+9",
              "sources": "/usr/local/jdk-11.0.21+9/lib/src.zip",
              "default": true
            },
            {
              "name": "JavaSE-17",
              "path": "/usr/lib/jvm/java-17-openjdk-amd64"
            }
          ]
        }
      }
    },
    "bundles": [
      "/home/huang/jdtls/com.microsoft.java.debug.plugin-0.51.1.jar"
    ]
  }
}
```
the configuration of `runtimes` works fine. But `autobuild` doesn't work.

**Comment by huangfeiyu** [comment]
Hi guys, any other suggestions on this? Without setting of "java.autobuild.enabled=false", it takes me 3 to 5 minutes to wait before I can edit using the functions provided by eglot.

**Comment by huangfeiyu** [comment]
Hi @joaotavora Can you please help me? I know you are busy, but this issue really bothers me. Thanks!

**Reply by joaotavora** [reply]
Sorry @huangfeiyu , I don't have time to install a Java toolkit in my machine.  Is it that `M-x eglot` takes a long time to run?  I don't understand the issue you are facing.  If the server takes a long time to respond synchronously to `M-x eglot`, maybe you can tweak `eglot-connect-timeout`.   I have no idea what "autobuild" is. (and I don't have time to learn it, sorry).

**Reply by huangfeiyu** [reply]
@joaotavora thank you for replying me. the slow start issue got mitigated after I changed my computer. but still the config in my previous comment to set java.autobuild.enabled=false does not work. Other settings such as java.completion.importOrder=["com", "org", "java"] (configured in emacs json format) don't work either. according emacs doc, those should work. this isn't an issue of jdtls, because the settings work in vscode as well as lsp-java.

**Reply by joaotavora** [reply]
> Other settings such as java.completion.importOrder=["com", "org", "java"] (configured in emacs json format) don't work either. according emacs doc, those should work. t

Is there an "emacs doc" about these java settings?

I don't have any Java expertise, I'm sorry. If jtdls works with other LSP clients, it's likely they are smarter and configuring it in a different manner via the many configuration mechanisms.  You have to ask some jdtls developer of _how_ exactly to configure it. Or an Emacs user that has solved this problem, like @yveszoundi (who has even created a separate package to use on top of Eglot -- though I don't support it you could maybe find some hint there).

**Reply by huangfeiyu** [reply, marked as answer]
@joaotavora I am glad to tell you that my issue has been solved. It is such an issue haunts me for months. The solution is to add 
```
(add-hook 'java-ts-mode-hook (lambda ()
    (remove-hook 'eglot-connect-hook 'eglot-signal-didChangeConfiguration)))
```
The solution I got from https://github.com/joaotavora/eglot/discussions/1222#discussioncomment-5871986 You guys discussed the problem in it thoroughly.
By the way, may be you made a mistake in the comment https://github.com/joaotavora/eglot/discussions/1222#discussioncomment-5871986 , the following code you proposed doesn't work
```
(add-hook 'java-mode-hook (lambda ()
   (remove-hook 'eglot-connect-hook #'eglot-signal-didChangeConfiguration t))
```

My complete eglot jdtls configuration is as the following now
```
(with-eval-after-load 'eglot
  (add-to-list 'eglot-server-programs
               `(java-ts-mode . ("/home/huang/jdtls/bin/jdtls" :initializationOptions
                                 (
                                  :bundles ["/home/huang/jdtls/com.microsoft.java.debug.plugin-0.51.1.jar"]
                                  :settings
                                  (
                                   :java
                                   (
                                    :configuration (:runtimes [(:name "JavaSE-11" :path "/usr/local/jdk-11.0.21+9" :sources "/usr/local/jdk-11.0.21+9/lib/src.zip" :default t)
                                                               (:name "JavaSE-17" :path "/usr/lib/jvm/java-17-openjdk-amd64")])
                                    :saveActions (:organizeImports t)
                                    :autobuild (:enabled :json-false)
                                    :completion (
                                                 :enabled t
                                                 :overwrite t
                                                 :guessMethodArguments t
                                                 :importOrder ["java", "javax", "jakarta", "org", "co", "com", "io", "net", "software"])
                                    )
                                   )
                                  )
                                 )))
  (add-hook 'java-ts-mode-hook (lambda ()
                                 (remove-hook 'eglot-connect-hook 'eglot-signal-didChangeConfiguration)))
  )
```
I hope the issue related to quirky jdtls can be documented somewhere that can be reached easily. To figure out the issue I followed your advise to visit @yveszoundi 's package eglot-java, and in https://github.com/yveszoundi/eglot-java/issues/44#issuecomment-1986672691 he mentioned the issue https://github.com/joaotavora/eglot/discussions/1222.
Thank you and all others involved.

**Reply by joaotavora** [reply]
I'm very glad you solved it.  I didn't understand most of your report, I have to say.  If I had understood it, I would have directed you to that comment.

**Reply by huangfeiyu** [reply]
That is all right. It's difficult to relate autobuild, importOrder to format template if you are not a Java developer.
